src/app/websockets/handler.rs

```diff
@@ -0,0 +1,120 @@
+            return Err(Error::from(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "Unauthorized",
+            )))
+        }
+    };
+
+    let conn_id = manager_handle
+        .connect(session_token.clone(), &session)
+        .await
+        .map_err(Error::from)?;
+
+    log::info!("WebSocket connected: Connection ID {:?}", conn_id);
+
+    let session_clone = session.clone();
+    actix_web::rt::spawn(async move {
+        let mut last_heartbeat = Instant::now();
+        let mut interval = interval(HEARTBEAT_INTERVAL);
+        let mut session = session_clone;
+
+        loop {
+            tokio::select! {
+                Some(msg) = msg_stream.next() => {
+                    match msg {
+                        Ok(msg) => {
+                            if handle_message(msg, &mut session, &manager_handle, conn_id, &session_token).await.is_err() {
+                                break;
+                            }
+                            last_heartbeat = Instant::now();
+                        }
+                        Err(e) => {
+                            log::error!("Error in websocket: {:?}", e);
+                            break;
+                        }
+                    }
+                }
+                _ = interval.tick() => {
+                    if Instant::now().duration_since(last_heartbeat) > CLIENT_TIMEOUT {
+                        log::info!("Client timeout: Connection ID {:?}", conn_id);
+                        break;
+                    }
+                    if session.ping(b"").await.is_err() {
+                        break;
+                    }
+                }
+                else => break,
+            }
+        }
+
+        log::info!("WebSocket connection closed: Connection ID {:?}", conn_id);
+        manager_handle.disconnect(conn_id).await.ok();
+    });
+
+    Ok(response)
+}
+
+async fn handle_message(
+    msg: Message,
+    session: &mut Session,
+    manager_handle: &WebSocketManagerHandle,
+    conn_id: ConnId,
+    session_token: &SessionToken,
+) -> Result<(), Error> {
+    match msg {
+        Message::Ping(bytes) => {
+            session
+                .pong(&bytes)
+                .await
+                .map_err(|e| Error::from(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
+        }
+        Message::Pong(_) => {
+            manager_handle.update_heartbeat(conn_id).await?;
+        }
+        Message::Text(text) => {
+            log::info!("Text message received: {:?}", text);
+            let message = Message::Text(text);
+            manager_handle
+                .send_message_to_session(session_token, message, conn_id)
+                .await?;
+        }
+        Message::Binary(bin) => {
+            log::info!("Binary message received: {:?}", bin);
+            let message = Message::Binary(bin);
+            manager_handle
+                .send_message_to_session(session_token, message, conn_id)
```

Comment:
Because of the way the `websocket_message_to_bytestring` [func](https://github.com/Extheoisah/hxckr-core/blob/ae3f14f680e39260139b543f790c0542a7f01b78/src/shared/utils.rs#L21) has been implemented if we pass a Binary type message to the `send_message_to_session` function it will be converted to an empty string.  I think we either need to handle binary type messages in `send_message_to_session` or perhaps we want to exclude them entirely if we know we will never use them on the client.  What do you think?

Reply:
One suggestion for a way to handle it in `send_message_to_session`:

```rust
pub async fn send_message_to_session(
    &self,
    session_token: &SessionToken,
    message: Message,
    sender_conn_id: ConnId,
) -> io::Result<()> {
    let sessions = self.sessions.read().await;
    let connections = self.connections.read().await;

    if let Some(conn_ids) = sessions.get(session_token) {
        for &conn_id in conn_ids {
            log::info!(
                "sender_conn_id: {:?}, conn_id: {:?}",
                sender_conn_id,
                conn_id
            );
            if let Some(conn) = connections.get(&conn_id) {
                let mut session = conn.session.clone();
                match message {
                    Message::Text(ref text) => {
                        if let Err(e) = session.text(text).await {
                            log::error!("Failed to send text message to connection {}: {:?}", conn_id, e);
                        }
                    },
                    Message::Binary(ref bytes) => {
                        if let Err(e) = session.binary(bytes.clone()).await {
                            log::error!("Failed to send binary message to connection {}: {:?}", conn_id, e);
                        }
                    },
                    Message::Ping(ref bytes) => {
                        if let Err(e) = session.ping(bytes.clone()).await {
                            log::error!("Failed to send ping message to connection {}: {:?}", conn_id, e);
                        }
                    },
                    Message::Pong(ref bytes) => {
                        if let Err(e) = session.pong(bytes.clone()).await {
                            log::error!("Failed to send pong message to connection {}: {:?}", conn_id, e);
                        }
                    },
                    Message::Close(reason) => {
                        if let Err(e) = session.close(reason).await {
                            log::error!("Failed to send close message to connection {}: {:?}", conn_id, e);
                        }
                    },
                    // Handle other message types if necessary
                }
            }
        }
    }
    Ok(())
}
```

Reply:
i don't think we'll need to send or receive binary messages as we'll be sending just JSON text to the client which was why I only handled the text case. 
But I have refactored it to account for binary messages. Its not robust because it doesn't account for complex binary decoding. Given we don't have a use case for it now, I think it's not a deal breaker.